Modules/_zoneinfo.c

```diff
@@ -2226,22 +2248,108 @@ ymd_to_ord(int y, int m, int d)
 {
     y -= 1;
     int days_before_year = (y * 365) + (y / 4) - (y / 100) + (y / 400);
-    int yearday = DAYS_BEFORE_MONTH[m];
-    if (m > 2 && is_leap_year(y + 1)) {
-        yearday += 1;
-    }
+    int yearday = days_before_month(y + 1, m);
 
     return days_before_year + yearday + d;
 }
 
+/* year, month -> number of days in that month in that year */
+static int
+days_in_month(int year, int month)
+{
+    assert(month >= 1);
+    assert(month <= 12);
+    if (month == 2 && is_leap_year(year))
+        return 29;
+    else
+        return DAYS_IN_MONTH[month];
+}
+
+/* year, month -> number of days in year preceding first day of month */
+static int
+days_before_month(int year, int month)
+{
+    int days;
+
+    assert(month >= 1);
+    assert(month <= 12);
+    days = DAYS_BEFORE_MONTH[month];
+    if (month > 2 && is_leap_year(year))
+        ++days;
+    return days;
+}
+
+/* Number of days in 4, 100, and 400 year cycles.  That these have
+ * the correct values is asserted in the module init function.
+ */
+#define DI4Y    1461    /* days_before_year(5); days in 4 years */
+#define DI100Y  36524   /* days_before_year(101); days in 100 years */
+#define DI400Y  146097  /* days_before_year(401); days in 400 years  */
+
+/* Calculates year, month and day from an ordinal. */
+static void
+ord_to_ymd(int ordinal, int *year, int *month, int *day)
+{
+    int y, m, d;
+    int n, n1, n4, n100, n400, leapyear, preceding;
+
+    assert(ordinal >= 1);
+    --ordinal;
+    n400 = ordinal / DI400Y;
+    n = ordinal % DI400Y;
+    y = n400 * 400 + 1;
+
+    n100 = n / DI100Y;
+    n = n % DI100Y;
+
+    n4 = n / DI4Y;
+    n = n % DI4Y;
+
+    n1 = n / 365;
+    n = n % 365;
+
+    y += n100 * 100 + n4 * 4 + n1;
+    if (n1 == 4 || n100 == 4) {
+        assert(n == 0);
+        y -= 1;
+        m = 12;
+        d = 31;
+        return;
+    }
+
+    leapyear = n1 == 3 && (n4 != 24 || n100 == 3);
+    assert(leapyear == is_leap_year(y));
+    m = (n + 50) >> 5;
+    preceding = (DAYS_BEFORE_MONTH[m] + (m > 2 && leapyear));
+    if (preceding > n) {
+        m -= 1;
+        preceding -= days_in_month(y, m);
+    }
+    n -= preceding;
+    assert(0 <= n);
+    assert(n < days_in_month(y, m));
+
+    d = n + 1;
+
+    if (year != NULL) {
+        *year = y;
+    }
+    if (month != NULL) {
+        *month = m;
+    }
+    if (day != NULL) {
+        *day = d;
+    }
```

Comment:
I really, *really* dislike single-letter names, especially in a codebase as large as this, and *especially* with a lot of mathematical operations.

Reply:
> I really, _really_ dislike single-letter names, especially in a codebase as large as this, and _especially_ with a lot of mathematical operations.

I should comment that this is copied from `_datetimemodule.c`

Reply:
P.S. The codebase in `_datetimemodule.c` and `_zoneinfo.c` are already a mess. If possible I could refactor them later in a new PR.

Reply:
Oh, it's copied. In that case, it might be better to just add an API function to `datetime` and call it.

Reply:
> Oh, it's copied. In that case, it might be better to just add an API function to `datetime` and call it.

Right now, these 2 modules have multiple almost-same temporal functions. They should have a reason to do so but I don't think here to add and call an API in `datetime` is clean enough.

Reply:
Why wouldn't it be "clean" enough?

Reply:
We probably have only two "clean" choices here, while please notice that I'm not saying the current draft is the clean one, instead it is quite hacky, ugly and breaks `fold` property (= 0). I will leave the details at the original issue.